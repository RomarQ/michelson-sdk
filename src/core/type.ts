import { MichelsonJSON, MichelsonMicheline, PairsOfKeys } from '../typings';
import { IType, PrimType } from '../typings/type';
import { Prim } from './enums/prim';

export class Michelson_Type implements IType {
    #annotation?: string;

    private innerTypes: IType[];

    constructor(private type: PrimType, ...innerTypes: IType[]) {
        this.innerTypes = innerTypes;
    }

    /**
     * @description Set field annotation
     * @link https://tezos.gitlab.io/active/michelson.html#field-and-constructor-annotations
     * @param {string} annotation field annotation
     */
    public setAnnotation(annotation: string) {
        this.#annotation = annotation;
        return this;
    }

    /**
     * @description Generate the Micheline representation of the type
     * @returns {MichelsonMicheline} Micheline representation
     */
    toMicheline(): MichelsonMicheline {
        const expr = this.#annotation ? `(${this.type} %${this.#annotation})` : `${this.type}`;
        switch (this.type) {
            case Prim.unit:
            case Prim.int:
            case Prim.nat:
            case Prim.mutez:
            case Prim.timestamp:
            case Prim.string:
            case Prim.address:
            case Prim.chain_id:
            case Prim.bool:
            case Prim.bytes:
            case Prim.bls12_381_fr:
            case Prim.bls12_381_g1:
            case Prim.bls12_381_g2:
            case Prim.key:
            case Prim.key_hash:
            case Prim.signature:
            case Prim.operation:
            case Prim.never:
                return expr;
            case Prim.list:
            case Prim.set:
            case Prim.pair:
            case Prim.or:
            case Prim.option:
            case Prim.map:
            case Prim.big_map:
            case Prim.lambda:
            case Prim.ticket:
            case Prim.contract:
            case Prim.sapling_state:
            case Prim.sapling_transaction:
                return `(${[expr, ...this.innerTypes.map((t) => t.toMicheline())].join(' ')})`;
        }
    }

    /**
     * @description Generate the JSON representation of the type
     * @returns {MichelsonMicheline} JSON representation
     */
    toJSON(): MichelsonJSON {
        const obj = this.#annotation ? { annots: [`%${this.#annotation}`] } : {};
        switch (this.type) {
            case Prim.unit:
            case Prim.int:
            case Prim.nat:
            case Prim.mutez:
            case Prim.timestamp:
            case Prim.string:
            case Prim.address:
            case Prim.chain_id:
            case Prim.bool:
            case Prim.bytes:
            case Prim.bls12_381_fr:
            case Prim.bls12_381_g1:
            case Prim.bls12_381_g2:
            case Prim.key:
            case Prim.key_hash:
            case Prim.signature:
            case Prim.operation:
            case Prim.never:
                return {
                    ...obj,
                    prim: this.type,
                };
            case Prim.list:
            case Prim.set:
            case Prim.pair:
            case Prim.or:
            case Prim.option:
            case Prim.map:
            case Prim.big_map:
            case Prim.lambda:
            case Prim.ticket:
            case Prim.contract:
            case Prim.sapling_state:
            case Prim.sapling_transaction:
                return {
                    ...obj,
                    prim: this.type,
                    args: this.innerTypes.map((t) => t.toJSON()),
                };
        }
    }

    /**
     * @description Resolve type instance to a primitive
     * @return {MichelsonJSON} Michelson JSON format
     */
    [Symbol.toPrimitive](): MichelsonJSON {
        return this.toJSON();
    }
}

export class Michelson_Type_With_Param implements IType {
    #annotation?: string;

    #params: number[];

    constructor(private type: PrimType, ...params: number[]) {
        this.#params = params;
    }

    /**
     * @description Set field annotation
     * @link https://tezos.gitlab.io/active/michelson.html#field-and-constructor-annotations
     * @param {string} annotation field annotation
     */
    public setAnnotation(annotation: string) {
        this.#annotation = annotation;
        return this;
    }

    /**
     * @description Generate the Micheline representation of the type
     * @returns {MichelsonMicheline} Micheline representation
     */
    toMicheline(): MichelsonMicheline {
        return `(${[this.type, ...this.#params.map(String)].join(' ')})`;
    }

    /**
     * @description Generate the JSON representation of the type
     * @returns {MichelsonMicheline} JSON representation
     */
    toJSON(): MichelsonJSON {
        const obj = this.#annotation ? { annots: [`%${this.#annotation}`] } : {};
        return {
            ...obj,
            prim: this.type,
            args: this.#params.map((p) => ({
                int: String(p),
            })),
        };
    }

    /**
     * @description Resolve type instance to a primitive
     * @return {MichelsonJSON} Michelson JSON format
     */
    [Symbol.toPrimitive](): MichelsonJSON {
        return this.toJSON();
    }
}

export class Michelson_Type_Record<T extends Record<string, IType> = Record<string, IType>> implements IType {
    _isType = true as const;
    #annotation?: string;
    #fields: T;
    // Default: right combs => https://tezos.gitlab.io/active/michelson.html#operations-on-pairs-and-right-combs
    #layout: PairsOfKeys<keyof T>;

    constructor(fields: T, layout?: PairsOfKeys<keyof T>) {
        this.#fields = fields;
        this.#layout =
            layout || Michelson_Type_Record.composeRightCombLayout<keyof T>(Object.keys(fields) as (keyof T)[]);
    }

    static composeRightCombLayout = <K>(fields: K[]): PairsOfKeys<K> => {
        if (fields.length > 2) {
            return [fields[0], this.composeRightCombLayout<K>(fields.slice(1))];
        }
        return fields;
    };

    /**
     * @description Set field annotation
     * @link https://tezos.gitlab.io/active/michelson.html#field-and-constructor-annotations
     * @param {string} annotation field annotation
     */
    public setAnnotation(annotation: string) {
        this.#annotation = annotation;
        return this;
    }

    /**
     * @description Generate the Micheline representation of the type
     * @param fields Record fields
     * @param layout Record layout
     * @returns {MichelsonMicheline} Micheline representation
     */
    private _toMicheline(fields: T, layout: PairsOfKeys<keyof T>): MichelsonMicheline {
        const annotation = this.#annotation ? ` %${this.#annotation}` : '';
        const innerTypes = layout
            .map((layout) => {
                if (Array.isArray(layout)) {
                    return this._toMicheline(fields, layout);
                }
                return fields[layout].toMicheline();
            }, '')
            .join(' ');
        return `(${Prim.pair}${annotation} ${innerTypes})`;
    }

    /**
     * @description Generate the Micheline representation of the type
     * @returns {MichelsonMicheline} Micheline representation
     */
    public toMicheline(): MichelsonMicheline {
        return this._toMicheline(this.#fields, this.#layout);
    }

    /**
     * @description Generate the JSON representation of the type
     * @param fields Record fields
     * @param layout Record layout
     * @returns {MichelsonMicheline} JSON representation
     */
    private _toJSON(fields: T, layout: PairsOfKeys<keyof T>): MichelsonJSON {
        return {
            prim: Prim.pair,
            ...(this.#annotation ? { annots: [`%${this.#annotation}`] } : {}),
            args: layout.map((layout) => {
                if (Array.isArray(layout)) {
                    return this._toJSON(fields, layout);
                }

                return fields[layout].toJSON();
            }, []),
        };
    }

    /**
     * @description Generate the JSON representation of the type
     * @returns {MichelsonMicheline} JSON representation
     */
    public toJSON(): MichelsonJSON {
        return this._toJSON(this.#fields, this.#layout);
    }

    /**
     * @description Resolve type instance to a primitive
     * @return {MichelsonJSON} Michelson JSON format
     */
    [Symbol.toPrimitive](): MichelsonJSON {
        return this.toJSON();
    }
}

// Singleton types
export const TNat = () => new Michelson_Type(Prim.nat);
export const TInt = () => new Michelson_Type(Prim.int);
export const TMutez = () => new Michelson_Type(Prim.mutez);
export const TString = () => new Michelson_Type(Prim.string);
export const TBool = () => new Michelson_Type(Prim.bool);
export const TAddress = () => new Michelson_Type(Prim.address);
export const TTimestamp = () => new Michelson_Type(Prim.timestamp);
export const TChain_id = () => new Michelson_Type(Prim.chain_id);
export const TBytes = () => new Michelson_Type(Prim.bytes);
export const TBls12_381_fr = () => new Michelson_Type(Prim.bls12_381_fr);
export const TBls12_381_g1 = () => new Michelson_Type(Prim.bls12_381_g1);
export const TBls12_381_g2 = () => new Michelson_Type(Prim.bls12_381_g2);
export const TKey = () => new Michelson_Type(Prim.key);
export const TKey_hash = () => new Michelson_Type(Prim.key_hash);
export const TSignature = () => new Michelson_Type(Prim.signature);
export const TUnit = () => new Michelson_Type(Prim.unit);
export const TOperation = () => new Michelson_Type(Prim.operation);
export const TNever = () => new Michelson_Type(Prim.never);
// Container types
export const TList = (innerType: IType) => new Michelson_Type(Prim.list, innerType);
export const TSet = (innerType: IType) => new Michelson_Type(Prim.set, innerType);
export const TOption = (innerType: IType) => new Michelson_Type(Prim.option, innerType);
export const TPair = (leftType: IType, rightType: IType) => new Michelson_Type(Prim.pair, leftType, rightType);
export const TOr = (leftType: IType, rightType: IType) => new Michelson_Type(Prim.or, leftType, rightType);
export const TMap = (keyType: IType, valueType: IType) => new Michelson_Type(Prim.map, keyType, valueType);
export const TBig_map = (keyType: IType, valueType: IType) => new Michelson_Type(Prim.big_map, keyType, valueType);
export const TLambda = (inType: IType, outType: IType) => new Michelson_Type(Prim.lambda, inType, outType);
export const TTicket = (innerType: IType) => new Michelson_Type(Prim.ticket, innerType);
export const TContract = (innerType: IType) => new Michelson_Type(Prim.contract, innerType);
export const TSapling_state = (memoSize: number) => new Michelson_Type_With_Param(Prim.sapling_state, memoSize);
export const TSapling_transaction = (memoSize: number) =>
    new Michelson_Type_With_Param(Prim.sapling_transaction, memoSize);
// Artificial Types
export const TRecord = (fields: Record<string, IType>, layout?: PairsOfKeys<string>) =>
    new Michelson_Type_Record(fields, layout);

const Types = {
    // Singleton types
    TUnit,
    TNat,
    TInt,
    TMutez,
    TString,
    TBool,
    TBytes,
    TAddress,
    TTimestamp,
    TChain_id,
    TBls12_381_fr,
    TBls12_381_g1,
    TBls12_381_g2,
    TKey,
    TKey_hash,
    TSignature,
    TOperation,
    TNever,
    // Container types
    TList,
    TSet,
    TOption,
    TPair,
    TOr,
    TMap,
    TBig_map,
    TLambda,
    TTicket,
    TContract,
    TSapling_state,
    TSapling_transaction,
    // Artificial Types
    TRecord,
};

export default Types;
