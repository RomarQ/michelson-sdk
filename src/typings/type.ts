export interface IType {
    _isType: true;
}
