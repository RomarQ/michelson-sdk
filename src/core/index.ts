export * from './type';
export * from './literal';

export * as Type from './type';
export * as Literal from './literal';
