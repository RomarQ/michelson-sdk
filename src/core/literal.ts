import { Michelson_Type_Record, TOption, TRecord } from './type';
import Utils from '../misc/utils';
import { MichelsonJSON, MichelsonMicheline } from '../typings';
import { ILayout, IType } from '../typings/type';
import { Prim } from './enums/prim';
import {
    Michelson_Type,
    TAddress,
    TBool,
    TBytes,
    TChainID,
    TInt,
    TList,
    TMutez,
    TNat,
    TString,
    TTimestamp,
    TUnit,
} from './type';

export type Michelson_LiteralUnion = Michelson_Literal | Michelson_Literal_C1;

export class Michelson_Literal {
    private prim: Prim;
    private value: number | string | boolean;
    type: Michelson_Type;

    constructor(prim: Prim, type: Michelson_Type, value?: number | string | boolean) {
        this.prim = prim;
        this.type = type;
        switch (prim) {
            case Prim.Unit:
            case Prim.None:
                this.value = prim;
                break;
            default:
                if (typeof value === 'undefined') {
                    throw new Error('Expected a value!');
                }
                this.value = value;
        }
    }

    toMicheline(): MichelsonMicheline {
        switch (this.prim) {
            case Prim.None:
            case Prim.Unit:
            case Prim.int:
            case Prim.nat:
            case Prim.mutez:
            case Prim.bytes:
                return `${this.value}`;
            case Prim.chain_id:
                if (`${this.value}`.slice(0, 2) === '0x') {
                    return `${this.value}`;
                }
                return `"${this.value}"`;
            case Prim.address:
            case Prim.string:
                return `"${this.value}"`;

            case Prim.timestamp:
                if (typeof this.value === 'string') {
                    return `"${this.value}"`;
                }
                return `${this.value}`;
            case Prim.bool:
                return Utils.capitalizeBoolean(this.value as boolean);
        }

        throw new Error(`Cannot produce michelson for literal of type: ${this.prim}`);
    }

    toJSON(): Record<string, unknown> {
        switch (this.prim) {
            case Prim.None:
            case Prim.Unit:
                return {
                    prim: this.value,
                };
            case Prim.int:
            case Prim.nat:
            case Prim.mutez:
                return {
                    [Prim.int]: `${this.value}`,
                };
            case Prim.string:
            case Prim.address:
                return {
                    [Prim.string]: this.value,
                };
            case Prim.bytes:
                return {
                    // Same behaviour as in "tezos-client"
                    [Prim.bytes]: Utils.compressHexString(`${this.value}`),
                };
            case Prim.chain_id:
                if (`${this.value}`.slice(0, 2) === '0x') {
                    return {
                        // Same behaviour as in "tezos-client"
                        [Prim.bytes]: Utils.compressHexString(`${this.value}`),
                    };
                }
                return {
                    [Prim.string]: this.value,
                };
            case Prim.timestamp:
                if (typeof this.value === 'string') {
                    return {
                        [Prim.string]: this.value,
                    };
                }
                return {
                    [Prim.int]: `${this.value}`,
                };
            case Prim.bool:
                return {
                    prim: Utils.capitalizeBoolean(this.value as boolean),
                };
        }

        throw new Error(`Cannot produce michelson JSON for literal of type: ${this.prim}`);
    }
}

export class Michelson_Literal_C1 {
    #prim: Prim;
    #elements: Michelson_LiteralUnion[];
    type: Michelson_Type;

    constructor(prim: Prim, type: Michelson_Type, elements: Michelson_LiteralUnion[]) {
        this.#prim = prim;
        this.type = type;
        this.#elements = elements;
    }

    toMicheline(): MichelsonMicheline {
        switch (this.#prim) {
            case Prim.Some:
                return `${this.#prim} ${this.#elements.map((v) => v.toMicheline()).join(' ; ')}`;
            case Prim.list:
                return `{ ${this.#elements.map((v) => v.toMicheline()).join(' ; ')} }`;
        }

        throw new Error(`Cannot produce michelson for literal of type: ${this.#prim}`);
    }

    toJSON(): MichelsonJSON {
        switch (this.#prim) {
            case Prim.Some:
                return {
                    prim: this.#prim,
                    args: this.#elements.map((v) => v.toJSON()),
                };
            case Prim.list:
                return this.#elements.map((v) => v.toJSON());
        }

        throw new Error(`Cannot produce michelson JSON for literal of type: ${this.#prim}`);
    }
}

class Michelson_Record {
    #fields: Record<string, Michelson_LiteralUnion>;
    type: IType;
    // Default layout => https://tezos.gitlab.io/active/michelson.html#operations-on-pairs-and-right-combs
    #layout: ILayout;

    constructor(fields: Record<string, Michelson_LiteralUnion>, layout?: ILayout) {
        this.#fields = fields;
        this.#layout = layout || Michelson_Type_Record.composeRightCombLayout(Object.keys(fields));
        this.type = TRecord(
            Object.entries(fields).reduce((pv, [key, value]) => {
                return {
                    ...pv,
                    [key]: value.type.setAnnotation(key),
                };
            }, {}),
            this.#layout,
        );
    }

    /**
     * @description Generate the Micheline representation
     * @param fields Record fields
     * @param layout Record layout
     * @returns {MichelsonMicheline} Micheline representation
     */
    private _toMicheline(
        fields: globalThis.Record<string, Michelson_LiteralUnion>,
        layout: ILayout,
    ): MichelsonMicheline {
        const innerTypes = layout
            .map((layout) => {
                if (Array.isArray(layout)) {
                    return this._toMicheline(fields, layout as ILayout);
                }

                return fields[layout].toMicheline();
            }, '')
            .join(' ');
        return `(${Prim.Pair} ${innerTypes})`;
    }

    /**
     * @description Generate the Micheline representation
     * @returns {MichelsonMicheline} Micheline representation
     */
    public toMicheline(): MichelsonMicheline {
        return this._toMicheline(this.#fields, this.#layout);
    }

    /**
     * @description Generate the JSON representation
     * @param fields Record fields
     * @param layout Record layout
     * @returns {MichelsonMicheline} JSON representation
     */
    private _toJSON(fields: Record<string, Michelson_LiteralUnion>, layout: ILayout): MichelsonJSON {
        return {
            prim: Prim.Pair,
            args: layout.map((layout) => {
                if (Array.isArray(layout)) {
                    return this._toJSON(fields, layout as ILayout);
                }

                return fields[layout].toJSON();
            }, []),
        };
    }

    /**
     * @description Generate the JSON representation
     * @returns {MichelsonMicheline} JSON representation
     */
    public toJSON(): MichelsonJSON {
        return this._toJSON(this.#fields, this.#layout);
    }
}

export const Unit = () => new Michelson_Literal(Prim.Unit, TUnit);

export const Nat = (value: number) => new Michelson_Literal(Prim.nat, TNat, value);
export const Int = (value: number) => new Michelson_Literal(Prim.int, TInt, value);
export const Mutez = (value: number) => new Michelson_Literal(Prim.mutez, TMutez, value);
export const Timestamp = (value: number | string) => new Michelson_Literal(Prim.timestamp, TTimestamp, value);

export const String = (value: string) => new Michelson_Literal(Prim.string, TString, value);
export const Address = (value: string) => new Michelson_Literal(Prim.address, TAddress, value);
export const Bytes = (value: string) => new Michelson_Literal(Prim.bytes, TBytes, value);
export const ChainID = (value: string) => new Michelson_Literal(Prim.chain_id, TChainID, value);

export const Bool = (value: boolean) => new Michelson_Literal(Prim.bool, TBool, value);

export const List = (elements: Michelson_LiteralUnion[], innerType: Michelson_Type) =>
    new Michelson_Literal_C1(Prim.list, TList(innerType), elements);

export const None = (innerType: Michelson_Type) => new Michelson_Literal(Prim.None, TOption(innerType));
export const Some = (element: Michelson_LiteralUnion) =>
    new Michelson_Literal_C1(Prim.Some, TOption(element.type), [element]);

export const Record = (fields: Record<string, Michelson_LiteralUnion>, layout?: ILayout) =>
    new Michelson_Record(fields, layout);

const Literals = {
    Unit,
    Nat,
    Int,
    Mutez,
    String,
    Bool,
    Address,
    Timestamp,
    ChainID,
    Bytes,
    List,
    None,
    Some,
    Record,
};

export default Literals;
