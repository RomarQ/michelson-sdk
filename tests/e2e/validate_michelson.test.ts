import path from 'path';
import { execSync } from 'child_process';
import {
    Address,
    Bool,
    Bytes,
    ChainID,
    Int,
    List,
    Mutez,
    Nat,
    None,
    Pair,
    Record,
    Some,
    String,
    Timestamp,
    Unit,
    Set,
} from '../../src/core/literal';
import { TNat } from '../../src/core/type';

const TEZOS_CLIENT_CMD = './tests/e2e/tezos-binaries/tezos-client';

const convertMichelsonToJSON = (value: string, type: string) =>
    execSync(
        `${path.resolve(
            TEZOS_CLIENT_CMD,
        )} --mode mockup --base-dir _build/mockup convert data ${value} from michelson to json --type "${type}"
`,
    ).toString('utf-8');

if (process.platform == 'linux') {
    beforeAll(() => {
        try {
            execSync(path.resolve('./tests/e2e/scripts/bootstrap-mockup.sh'));
            console.info('Mockup bootstrapped...');
        } catch (e) {
            console.error(e);
        }
    });
    afterAll(() => {
        try {
            execSync(path.resolve('./tests/e2e/scripts/teardown.sh'));
            console.info('Teardown...');
        } catch (e) {
            console.error(e);
        }
    });

    describe('[E2E] - Michelson compilation (Singleton Literals)', () => {
        it('Nat', () => {
            const literal = Nat(1);
            const value = literal.toMicheline();
            const type = literal.type.toMicheline();
            const jsonValue = literal.toJSON();

            const result = convertMichelsonToJSON(value, type);

            expect(jsonValue).toEqual(JSON.parse(result));
            expect(jsonValue).toMatchSnapshot();
        });
        it('Int', () => {
            const literal = Int(2);
            const value = literal.toMicheline();
            const type = literal.type.toMicheline();
            const jsonValue = literal.toJSON();

            const result = convertMichelsonToJSON(value, type);

            expect(jsonValue).toEqual(JSON.parse(result));
            expect(jsonValue).toMatchSnapshot();
        });
        it('Mutez', () => {
            const literal = Mutez(10);
            const value = literal.toMicheline();
            const type = literal.type.toMicheline();
            const jsonValue = literal.toJSON();

            const result = convertMichelsonToJSON(value, type);

            expect(jsonValue).toEqual(JSON.parse(result));
            expect(jsonValue).toMatchSnapshot();
        });
        it('String', () => {
            const literal = String('Hello World');
            const value = literal.toMicheline();
            const type = literal.type.toMicheline();
            const jsonValue = literal.toJSON();

            const result = convertMichelsonToJSON(`'${value}'`, type);

            expect(jsonValue).toEqual(JSON.parse(result));
            expect(jsonValue).toMatchSnapshot();
        });
        it('Bool (True)', () => {
            const literal = Bool(true);
            const value = literal.toMicheline();
            const type = literal.type.toMicheline();
            const jsonValue = literal.toJSON();

            const result = convertMichelsonToJSON(value, type);

            expect(jsonValue).toEqual(JSON.parse(result));
            expect(jsonValue).toMatchSnapshot();
        });
        it('Bool (False)', () => {
            const literal = Bool(false);
            const value = literal.toMicheline();
            const type = literal.type.toMicheline();
            const jsonValue = literal.toJSON();

            const result = convertMichelsonToJSON(value, type);

            expect(jsonValue).toEqual(JSON.parse(result));
            expect(jsonValue).toMatchSnapshot();
        });
        it('Bytes', () => {
            const literal = Bytes('0xFFFF');
            const value = literal.toMicheline();
            const type = literal.type.toMicheline();
            const jsonValue = literal.toJSON();

            const result = convertMichelsonToJSON(value, type);

            expect(jsonValue).toEqual(JSON.parse(result));
            expect(jsonValue).toMatchSnapshot();
        });
        it('Address', () => {
            const literal = Address('tz28QJHLyqvaY2rXAoFZTbxrXeD88NA8wscC');
            const value = literal.toMicheline();
            const type = literal.type.toMicheline();
            const jsonValue = literal.toJSON();

            const result = convertMichelsonToJSON(`'${value}'`, type);

            expect(jsonValue).toEqual(JSON.parse(result));
            expect(jsonValue).toMatchSnapshot();
        });
        it('Timestamp (string)', () => {
            const literal = Timestamp('2019-09-26T10:59:51Z');
            const value = literal.toMicheline();
            const type = literal.type.toMicheline();
            const jsonValue = literal.toJSON();

            const result = convertMichelsonToJSON(`'${value}'`, type);

            expect(jsonValue).toEqual(JSON.parse(result));
            expect(jsonValue).toMatchSnapshot();
        });
        it('Timestamp (number)', () => {
            const literal = Timestamp(1571659294);
            const value = literal.toMicheline();
            const type = literal.type.toMicheline();
            const jsonValue = literal.toJSON();

            const result = convertMichelsonToJSON(value, type);

            expect(jsonValue).toEqual(JSON.parse(result));
            expect(jsonValue).toMatchSnapshot();
        });
        it('ChainID (string)', () => {
            const literal = ChainID('NetXynUjJNZm7wi');
            const value = literal.toMicheline();
            const type = literal.type.toMicheline();
            const jsonValue = literal.toJSON();

            const result = convertMichelsonToJSON(`'${value}'`, type);

            expect(jsonValue).toEqual(JSON.parse(result));
            expect(jsonValue).toMatchSnapshot();
        });
        it('ChainID (bytes)', () => {
            const literal = ChainID('0x7a06a770');
            const value = literal.toMicheline();
            const type = literal.type.toMicheline();
            const jsonValue = literal.toJSON();

            const result = convertMichelsonToJSON(value, type);

            expect(jsonValue).toEqual(JSON.parse(result));
            expect(jsonValue).toMatchSnapshot();
        });
        it('Unit', () => {
            const literal = Unit();
            const value = literal.toMicheline();
            const type = literal.type.toMicheline();
            const jsonValue = literal.toJSON();

            const result = convertMichelsonToJSON(`'${value}'`, type);

            expect(jsonValue).toEqual(JSON.parse(result));
            expect(jsonValue).toMatchSnapshot();
        });
    });

    describe('[E2E] - Michelson compilation (List)', () => {
        it('List', () => {
            const literal = List([Nat(1), Nat(2)], TNat);
            const value = literal.toMicheline();
            const type = literal.type.toMicheline();
            const jsonValue = literal.toJSON();

            const result = convertMichelsonToJSON(`"${value}"`, type);

            expect(jsonValue).toEqual(JSON.parse(result));
            expect(jsonValue).toMatchSnapshot();
        });
        it('Set', () => {
            const literal = Set([Nat(1), Nat(2)], TNat);
            const value = literal.toMicheline();
            const type = literal.type.toMicheline();
            const jsonValue = literal.toJSON();

            const result = convertMichelsonToJSON(`"${value}"`, type);

            expect(jsonValue).toEqual(JSON.parse(result));
            expect(jsonValue).toMatchSnapshot();
        });
    });

    describe('[E2E] - Michelson compilation (Option)', () => {
        it('Some', () => {
            const literal = Some(Nat(1));
            const value = literal.toMicheline();
            const type = literal.type.toMicheline();
            const jsonValue = literal.toJSON();

            const result = convertMichelsonToJSON(`'${value}'`, type);

            expect(jsonValue).toEqual(JSON.parse(result));
            expect(jsonValue).toMatchSnapshot();
        });
        it('None', () => {
            const literal = None(TNat);
            const value = literal.toMicheline();
            const type = literal.type.toMicheline();
            const jsonValue = literal.toJSON();

            const result = convertMichelsonToJSON(`'${value}'`, type);

            expect(jsonValue).toEqual(JSON.parse(result));
            expect(jsonValue).toMatchSnapshot();
        });
    });

    describe('[E2E] - Michelson compilation (Pair)', () => {
        it('Simple', () => {
            const literal = Pair(Nat(1), Unit());

            const value = literal.toMicheline();
            const type = literal.type.toMicheline();
            const jsonValue = literal.toJSON();

            const result = convertMichelsonToJSON(`'${value}'`, type);
            expect(jsonValue).toEqual(JSON.parse(result));
            expect(jsonValue).toMatchSnapshot();
        });
        it('Nested', () => {
            const literal = Pair(Nat(1), Pair(Bool(false), String('Test')));

            const value = literal.toMicheline();
            const type = literal.type.toMicheline();
            const jsonValue = literal.toJSON();

            const result = convertMichelsonToJSON(`'${value}'`, type);
            expect(jsonValue).toEqual(JSON.parse(result));
            expect(jsonValue).toMatchSnapshot();
        });
    });

    describe('[E2E] - Michelson compilation (Record)', () => {
        it('Left Comb Layout', () => {
            const literal = Record(
                {
                    field1: Nat(1),
                    field2: Nat(2),
                    field3: Nat(3),
                },
                [['field1', 'field2'], 'field3'],
            );
            const value = literal.toMicheline();
            const type = literal.type.toMicheline();
            const jsonValue = literal.toJSON();

            const result = convertMichelsonToJSON(`'${value}'`, type);
            expect(jsonValue).toEqual(JSON.parse(result));
            expect(jsonValue).toMatchSnapshot();
        });
        it('Balanced Layout', () => {
            const literal = Record(
                {
                    field1: Nat(1),
                    field2: Nat(2),
                    field3: Nat(3),
                    field4: Nat(4),
                },
                [
                    ['field1', 'field2'],
                    ['field3', 'field4'],
                ],
            );
            const value = literal.toMicheline();
            const type = literal.type.toMicheline();
            const jsonValue = literal.toJSON();

            const result = convertMichelsonToJSON(`'${value}'`, type);
            expect(jsonValue).toEqual(JSON.parse(result));
            expect(jsonValue).toMatchSnapshot();
        });
        it('Right Comb Layout', () => {
            const literal = Record({
                field1: Nat(1),
                field2: Nat(2),
                field3: Nat(3),
            });
            const value = literal.toMicheline();
            const type = literal.type.toMicheline();
            const jsonValue = literal.toJSON();

            const result = convertMichelsonToJSON(`'${value}'`, type);
            expect(jsonValue).toEqual(JSON.parse(result));
            expect(jsonValue).toMatchSnapshot();
        });
    });
}
