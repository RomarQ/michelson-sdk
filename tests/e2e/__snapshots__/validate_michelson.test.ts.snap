// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[E2E] - Michelson compilation (List) List 1`] = `
Array [
  Object {
    "int": "1",
  },
  Object {
    "int": "2",
  },
]
`;

exports[`[E2E] - Michelson compilation (List) Set 1`] = `
Array [
  Object {
    "int": "1",
  },
  Object {
    "int": "2",
  },
]
`;

exports[`[E2E] - Michelson compilation (Option) None 1`] = `
Object {
  "prim": "None",
}
`;

exports[`[E2E] - Michelson compilation (Option) Some 1`] = `
Object {
  "args": Array [
    Object {
      "int": "1",
    },
  ],
  "prim": "Some",
}
`;

exports[`[E2E] - Michelson compilation (Pair) Nested 1`] = `
Object {
  "args": Array [
    Object {
      "int": "1",
    },
    Object {
      "args": Array [
        Object {
          "prim": "False",
        },
        Object {
          "string": "Test",
        },
      ],
      "prim": "Pair",
    },
  ],
  "prim": "Pair",
}
`;

exports[`[E2E] - Michelson compilation (Pair) Simple 1`] = `
Object {
  "args": Array [
    Object {
      "int": "1",
    },
    Object {
      "prim": "Unit",
    },
  ],
  "prim": "Pair",
}
`;

exports[`[E2E] - Michelson compilation (Record) Balanced Layout 1`] = `
Object {
  "args": Array [
    Object {
      "args": Array [
        Object {
          "int": "1",
        },
        Object {
          "int": "2",
        },
      ],
      "prim": "Pair",
    },
    Object {
      "args": Array [
        Object {
          "int": "3",
        },
        Object {
          "int": "4",
        },
      ],
      "prim": "Pair",
    },
  ],
  "prim": "Pair",
}
`;

exports[`[E2E] - Michelson compilation (Record) Left Comb Layout 1`] = `
Object {
  "args": Array [
    Object {
      "args": Array [
        Object {
          "int": "1",
        },
        Object {
          "int": "2",
        },
      ],
      "prim": "Pair",
    },
    Object {
      "int": "3",
    },
  ],
  "prim": "Pair",
}
`;

exports[`[E2E] - Michelson compilation (Record) Right Comb Layout 1`] = `
Object {
  "args": Array [
    Object {
      "int": "1",
    },
    Object {
      "args": Array [
        Object {
          "int": "2",
        },
        Object {
          "int": "3",
        },
      ],
      "prim": "Pair",
    },
  ],
  "prim": "Pair",
}
`;

exports[`[E2E] - Michelson compilation (Singleton Literals) Address 1`] = `
Object {
  "string": "tz28QJHLyqvaY2rXAoFZTbxrXeD88NA8wscC",
}
`;

exports[`[E2E] - Michelson compilation (Singleton Literals) Bool (False) 1`] = `
Object {
  "prim": "False",
}
`;

exports[`[E2E] - Michelson compilation (Singleton Literals) Bool (True) 1`] = `
Object {
  "prim": "True",
}
`;

exports[`[E2E] - Michelson compilation (Singleton Literals) Bytes 1`] = `
Object {
  "bytes": "ffff",
}
`;

exports[`[E2E] - Michelson compilation (Singleton Literals) ChainID (bytes) 1`] = `
Object {
  "bytes": "7a06a770",
}
`;

exports[`[E2E] - Michelson compilation (Singleton Literals) ChainID (string) 1`] = `
Object {
  "string": "NetXynUjJNZm7wi",
}
`;

exports[`[E2E] - Michelson compilation (Singleton Literals) Int 1`] = `
Object {
  "int": "2",
}
`;

exports[`[E2E] - Michelson compilation (Singleton Literals) Mutez 1`] = `
Object {
  "int": "10",
}
`;

exports[`[E2E] - Michelson compilation (Singleton Literals) Nat 1`] = `
Object {
  "int": "1",
}
`;

exports[`[E2E] - Michelson compilation (Singleton Literals) String 1`] = `
Object {
  "string": "Hello World",
}
`;

exports[`[E2E] - Michelson compilation (Singleton Literals) Timestamp (number) 1`] = `
Object {
  "int": "1571659294",
}
`;

exports[`[E2E] - Michelson compilation (Singleton Literals) Timestamp (string) 1`] = `
Object {
  "string": "2019-09-26T10:59:51Z",
}
`;

exports[`[E2E] - Michelson compilation (Singleton Literals) Unit 1`] = `
Object {
  "prim": "Unit",
}
`;
